daemon/stop.go

```diff
@@ -38,52 +38,56 @@ func (daemon *Daemon) ContainerStop(name string, timeout *int) error {
+	// TODO propogate a context down to this function
+	ctx := context.TODO()
 	if !container.IsRunning() {
 		return nil
 	}
-
+	var wait time.Duration
+	if seconds >= 0 {
+		wait = time.Duration(seconds) * time.Second
+	}
+	success := func() error {
+		daemon.LogContainerEvent(container, "stop")
+		return nil
+	}
 	stopSignal := container.StopSignal()
-	// 1. Send a stop signal
-	if err := daemon.killPossiblyDeadProcess(container, stopSignal); err != nil {
-		// While normally we might "return err" here we're not going to
-		// because if we can't stop the container by this point then
-		// it's probably because it's already stopped. Meaning, between
-		// the time of the IsRunning() call above and now it stopped.
-		// Also, since the err return will be environment specific we can't
-		// look for any particular (common) error that would indicate
-		// that the process is already dead vs something else going wrong.
-		// So, instead we'll give it up to 2 more seconds to complete and if
-		// by that time the container is still running, then the error
-		// we got is probably valid and so we force kill it.
-		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
-		defer cancel()
 
-		if status := <-container.Wait(ctx, containerpkg.WaitConditionNotRunning); status.Err() != nil {
-			logrus.Infof("Container failed to stop after sending signal %d to the process, force killing", stopSignal)
-			if err := daemon.killPossiblyDeadProcess(container, 9); err != nil {
-				return err
-			}
-		}
+	// 1. Send a stop signal
+	err := daemon.killPossiblyDeadProcess(container, stopSignal)
+	if err != nil {
+		wait = 2 * time.Second
 	}
 
-	// 2. Wait for the process to exit on its own
-	ctx := context.Background()
+	subCtx, cancel := context.WithCancel(ctx)
 	if seconds >= 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, time.Duration(seconds)*time.Second)
-		defer cancel()
+		// if seconds has been set, then set a timeout on the context so we dont wait forever
+		subCtx, cancel = context.WithTimeout(ctx, wait)
+	}
+	defer cancel()
+	if status := <-container.Wait(subCtx, containerpkg.WaitConditionNotRunning); status.Err() == nil {
+		// container did exit, so ignore any previous errors and return
+		return success()
 	}
 
-	if status := <-container.Wait(ctx, containerpkg.WaitConditionNotRunning); status.Err() != nil {
-		logrus.Infof("Container %v failed to exit within %d seconds of signal %d - using the force", container.ID, seconds, stopSignal)
-		// 3. If it doesn't, then send SIGKILL
-		if err := daemon.Kill(container); err != nil {
-			// Wait without a timeout, ignore result.
-			<-container.Wait(context.Background(), containerpkg.WaitConditionNotRunning)
-			logrus.Warn(err) // Don't return error because we only care that container is stopped, not what function stopped it
+	if err != nil {
+		// the container has still not exited, and the kill function errored, so log the error here:
```

Comment:
I think it's worth inverting the conditional and out-denting the logic here since there's nothing left to do if there was no error

```suggestion
	if err == nil {
		return success()
	}

	// the container has still not exited, and the kill function errored, so log the error here:
```

Reply:
sure i can do that

Reply:
I agree, but I also feel like there's an edge case here when the user specifies `seconds < 0` (thus no timeout), but the "wait for container to stop" context gets cancelled for other reasons, such as the client disconnecting.

My interpretation of `seconds < 0` is "client requested we wait forever" but that implies "client requested we 'stop' but never kill" so I think after the `if err == nil` block, we probably should also handle that edge case, right?

```suggestion
	if err == nil {
		return success()
	}
	if seconds < 0 {
		// if the client requested that we never kill / wait forever, but container.Wait was still interrupted (parent context cancelled, for example), we should propagate the signal failure
		return err
	}
	// the container has still not exited, and the kill function errored, so log the error here:
```

(Have to admit I'm a little bit out of my depth, so I might be misunderstanding the flow.)

Should we also explicitly invoke `cancel()` here since we know for sure that the current `subCtx` is finished? (before we confusingly reuse/shadow those variable names later in the "kill" block)